### ISSUE TITLE
How to get query_hash?

### ISSUE THREAD

**Issue by Kingson** [mineur/instagram-parser#7]
I would like to get query hash, any helpful info for me, thanks.

**Comment by alexhoma** [comment]
Hi! Here you have the specification: [/docs/setup.md#how-to-get-your-query-hash-old-query-id](https://github.com/mineur/instagram-parser/blob/master/docs/setup.md#how-to-get-your-query-hash-old-query-id)

Into the `docs/` directory you will see some examples and specifications of the library.

**Comment by Kingson** [comment]
@alexhoma Thanks for your replay, I read the docs, but I want to generate query hash by myself.
Instagram how to generate "42323d64886122307be10013ad2dcc44"?

**Comment by Kingson** [comment]
@alexhoma  query_hash does not change forever?

**Comment by idkm23** [comment]
what is the point of query_hash? I would expect it would be a way from instagram to confirm the query is coming from an informed sender. But if the hash is consistent, (i.e. doesn't have a parameter like time of day) then what is the point?

**Comment by hexadeciman** [comment]
> I would expect it would be a way from instagram to confirm the query is coming from an informed sender.

It is not, it's just a way for them to cache the query params so that they can store and return the response faster.
Read more here: https://stackoverflow.com/questions/54238696/what-is-query-hash-in-instagram

You should be able to generate it, using i.e. hash.js package

**Comment by jonbesga** [comment]
Hi @hexadeciman. I don't think the answer in the Stack Overflow thread is correct as the query_hash is always the same for every request (`f2405b236d85e8296cf30347c9f08c2a`) but the query params are differents everytime. If they were caching the query params with this method the query_hash should be different for each request.

**Comment by hexadeciman** [comment]
I just checked and it is different for every query 
![image](https://user-images.githubusercontent.com/5367205/61205150-99d86100-a6ef-11e9-8f68-621d0b4ed9fe.png)
It just seems to stay the same for pagination.

**Comment by jonbesga** [comment]
Hmm, weird.

For example, I'm loading [@mileycyrus](https://www.instagram.com/mileycyrus/) and [@instagram](https://www.instagram.com/instagram/).

First request for both accounts has the following query hashes (the same):
```
Miley: aec5501414615eca36a9acf075655b1e
Instagram: aec5501414615eca36a9acf075655b1e
```
In successive requests, for pagination, again the same:
```
Miley: query_hash: f2405b236d85e8296cf30347c9f08c2a
Instagram: query_hash: f2405b236d85e8296cf30347c9f08c2a
```

Also, they send all the data instead of first the query_hash. So this defeats the concept of the automated persisted queries

**Edit**: All of this is anonymously. I'm not logged in.

**Comment by oleksiivykhor** [comment]
Hello, @Kingson, query hash you can find in browser dev tools (https://github.com/mineur/instagram-parser/blob/master/docs/setup.md#how-to-get-your-query-hash-old-query-id) and then in neighbor tab JS you can find it in one of many loaded JS files, by Ctrl + F. Then before each request you can get content of this JS file from the link in html and parse generated query hash by regex.

**Comment by webszy** [comment]
Need help for stories?how can i get stories now?

**Comment by benocd** [comment]
Does the query hash ever expire? or does it have a query limit?

**Comment by KaKi87** [comment]
Hello,
So there is no way to automatically get the query hash ?

**Comment by meshchaninov** [comment]
query_hash hardcoded in one of *.js files on instagram. If u want to get hash, u should parse js files.

**Comment by KaKi87** [comment]
Yes, [I actually found out](https://git.kaki87.net/KaKi87/ig-scraper/src/branch/master/index.js#L190). Thanks !

**Comment by anuuzsoni** [comment]
> Yes, [I actually found out](https://git.kaki87.net/KaKi87/ig-scraper/src/branch/master/index.js#L190). Thanks !

how?? plzz help... I am stuck in a long time active project...plz help as soon as possble...thanku

**Comment by KaKi87** [comment]
Have a look at [this](https://git.kaki87.net/KaKi87/ig-scraper/src/branch/master/index.js#L304).

**Comment by anuuzsoni** [comment]
I did not understand...
and i have to do it in python...please guide..plz

**Comment by anuuzsoni** [comment]
> Have a look at [this](https://git.kaki87.net/KaKi87/ig-scraper/src/branch/master/index.js#L304).
i have to learn to create a query to get followers from end_cursor value that we get in the first and the following page
plz help me out

**Comment by hoosnick** [comment]
@meshchaninov thanks, really it is in the static JS files. 
Exactly the path: `/static/bundles/metro/ConsumerLibCommons.js/gEnEraTeD_sTr.js`

![instagram](https://user-images.githubusercontent.com/73847672/158746848-e8f1443b-e522-45a4-8da4-1add60583a46.png)

**Comment by Changlon** [comment]
> @meshchaninov谢谢，真的是在静态 JS 文件中。 具体路径：`/static/bundles/metro/ConsumerLibCommons.js/gEnEraTeD_sTr.js`
> 
> ![Instagram](https://user-images.githubusercontent.com/73847672/158746848-e8f1443b-e522-45a4-8da4-1add60583a46.png)

i do not kown why my queryId as same as yourself ! but the  name of  the .js filepath was different, and i can not access your .js filepath url

**Comment by thethiny** [comment]
hash_query does not hash the variables or the parameters, it hashes the path.
Lets say your actual path is `/graphql` and your payload is
```
{
    "user": {
        "profile": [
            "username",
            "user_id",
            "profile_picture"
        ],
        "feed": {
            "posts": {
                "data": [
                    "image_url"
                ],
                "page_size": "{{variables.max_count}}"
            }
        }
    }
}
```

Then this graphql payload will be hashed and it becomes `d4d88dc1500312af6f937f7b804c68c3`. Now instead of doing that on `/graphql` you do `/graphql/query/?query_hash=d4d88dc1500312af6f937f7b804c68c3`. This way you hashed the payload, as in you hashed the "keys" that are required from the graphql. So when you pass `variables` as a `param` then the payload does not actually change, because the variables are constant as well, and you are changing them on the backend, and not in the payload.

**Comment by Skuxblan** [comment]
Any thoughts how to get query hash now days? Seems it changed and we can't get it in same way anymore

**Comment by prinsi797** [comment]
you know what i am add this query_hash = "d4d88dc1500312af6f937f7b804c68c3" but not get data in Instagram reels 
add this values 
$variables = '{"child_comment_count":3,"fetch_comment_count":40,"has_threaded_comments":true,"parent_comment_count":24,"shortcode":"' . $reelId . '"}'; 

can you help how to get a query_hash value
